<b> Pandas Homework

In [1]:
# Dependencies
import pandas as pd

In [2]:
# Files to load
school_data_csv = 'Resources/schools_complete.csv'
student_data_csv = 'Resources/students_complete.csv'

# Read in data from csv files and store into DataFrames
school_data = pd.read_csv(school_data_csv)
student_data = pd.read_csv(student_data_csv)

In [3]:
# Evaluate school DataFrame head
school_data.head()

School ID            school_name      type  size   budget
0          0      Huang High School  District  2917  1910635
1          1   Figueroa High School  District  2949  1884411
2          2    Shelton High School   Charter  1761  1056600
3          3  Hernandez High School  District  4635  3022020
4          4    Griffin High School   Charter  1468   917500

In [4]:
# Evaluate student DataFrame head
student_data.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  
0             66          79  
1             94          61  
2             90          60  
3             67          58  
4             97          84

In [5]:
# Evaluate datatypes of student Dataframe
student_data.dtypes

Student ID        int64
student_name     object
gender           object
grade            object
school_name      object
reading_score     int64
math_score        int64
dtype: object

In [6]:
# Evaluate datatypes of school DataFrame
school_data.dtypes

School ID       int64
school_name    object
type           object
size            int64
budget          int64
dtype: object

In [7]:
# Add a column to the school_data DataFrame with the budget_per_student
school_data['budget_per_student'] = school_data['budget'] / school_data['size']

#school_data.head()

In [8]:
# Combine the data from the student and school DataFrames into a single DataFrame.  
school_data_complete = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])

school_data_complete.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  \
0             66          79          0  District  2917  1910635   
1             94          61          0  District  2917  1910635   
2             90          60          0  District  2917  1910635   
3             67          58          0  District  2917  1910635   
4             97          84          0  District  2917  1910635   

   budget_per_student  
0               655.0  
1               655.0  
2               655.0  
3               655.0  
4               655.0

In [9]:
# District Summary Table
# Find values to fill table
total_schools = school_data_complete['school_name'].nunique()
total_students = school_data_complete['Student ID'].nunique()
total_budget = school_data['budget'].sum()
average_math_score = school_data_complete['math_score'].mean()
average_reading_score = school_data_complete['reading_score'].mean()

In [10]:
# Additional DataFrames needed to find values for % passing
passing_math = school_data_complete.loc[school_data_complete['math_score'] >= 70, :]
passing_reading = school_data_complete.loc[school_data_complete['reading_score'] >= 70, :]
passing_both = school_data_complete.loc[(school_data_complete['math_score'] >= 70) & (student_data['reading_score'] >= 70), :]

# Find values to fill rest of District Summary Tables                          
percent_math = passing_math['Student ID'].nunique() / total_students * 100
percent_reading = passing_reading['Student ID'].nunique() / total_students * 100
percent_both = passing_both['Student ID'].nunique() / total_students * 100

<b>District Summary
* Calculate the total number of schools
* Calculate the total number of students
* Calculate the total budget
* Calculate the average math score
* Calculate the average reading score
* Calculate the percentage of students with a passing math score (70 or greater)
* Calculate the percentage of students with a passing reading score (70 or greater)
* Calculate the percentage of students who passed math and reading (% Overall Passing)
* Create a dataframe to hold the above results
* Optional: give the displayed data cleaner formatting

In [11]:
# Create District Summary Table
district_summary = pd.DataFrame({'Total Schools': [total_schools],
                                 'Total Students': total_students,
                                 'Total Budget': total_budget,
                                 'Average Math Score': average_math_score,
                                 'Average Reading Score': average_reading_score,
                                 '% Passing Math': percent_math,
                                 '% Passing Reading': percent_reading,
                                 '% Overall Passing': percent_both})
#district_summary

In [12]:
# Format to make District Summary Table more readable and accurately present data

# Add a thousand separator comma to Total Students column
district_summary['Total Students'] = district_summary['Total Students'].map("{:,}".format)

# Add a thousand separator comma, dollar sign, and decimal values to represents cents to Total Budget column
district_summary['Total Budget'] = district_summary['Total Budget'].map("${:,.2f}".format)

# Only show two decimal places on Average Math Score and Average Reading Score columns
district_summary['Average Math Score'] = district_summary['Average Math Score'].map("{:.2f}".format)
district_summary['Average Reading Score'] = district_summary['Average Reading Score'].map("{:.2f}".format)

# Only show two decimal places and show percent symbol on Percent Passing columns
district_summary['% Passing Math'] = district_summary['% Passing Math'].map("{:.2f}%".format)
district_summary['% Passing Reading'] = district_summary['% Passing Reading'].map("{:.2f}%".format)
district_summary['% Overall Passing'] = district_summary['% Overall Passing'].map("{:.2f}%".format)

district_summary

Total Schools Total Students    Total Budget Average Math Score  \
0             15         39,170  $24,649,428.00              78.99   

  Average Reading Score % Passing Math % Passing Reading % Overall Passing  
0                 81.88         74.98%            85.81%            65.17%

In [13]:
# School Summary Table
# store list of schools
schools = school_data_complete['school_name'].unique()

#schools

In [14]:
# GroupBy of merged DataFrame
student_by_school = school_data_complete.groupby(['School ID'])

# Use mean() to find average reading and math scores by school
school_average_math = student_by_school['math_score'].mean()
school_average_reading = student_by_school['reading_score'].mean()

In [15]:
# Store total number of students by school in a list
school_total_students = student_by_school['Student ID'].count()

#school_total_students

In [16]:
# GroupBy to filter the list of students passing math by school
passing_math_by_school = passing_math.groupby(['School ID'])

# Store total number of students passing math by school in a list
total_math_by_school = passing_math_by_school['Student ID'].count()

#total_math_by_school

In [17]:
# GroupBy to filter the list of students passing reading by school
passing_reading_by_school = passing_reading.groupby(['School ID'])

# Store total number of students passing reading by school in a list
total_reading_by_school = passing_reading_by_school['Student ID'].count()

#total_reading_by_school

In [18]:
# GroupBy to filter the list of students passing both math and reading by school
passing_both_by_school = passing_both.groupby(['School ID'])

# Store total number of students passing both math and reading by school in a list
total_both_by_school = passing_both_by_school['Student ID'].count()

#total_both_by_school

In [19]:
# Calculate % of students passing math by school
percent_math_by_school = total_math_by_school / school_total_students * 100

#percent_math_by_school

In [20]:
# Calculate % of students passing reading by school
percent_reading_by_school = total_reading_by_school / school_total_students * 100

#percent_reading_by_school

In [21]:
# Calculate % of students passing both math and reading by school
percent_both_by_school = total_both_by_school / school_total_students * 100

#percent_both_by_school

In [22]:
# Assign variable for budget_by_school
budget_by_school = school_data['budget']

#budget_by_school

In [23]:
# Assign variable for budget per student by school
budget_per_student_by_school = school_data['budget_per_student']

In [24]:
# Assign variable for school type
type_of_school = school_data['type']

<b>School Summary
* Create an overview table that summarizes key metrics about each school, including:
    * School Name
    * School Type
    * Total Students
    * Total School Budget
    * Per Student Budget
    * Average Math Score
    * Average Reading Score
    * % Passing Math
    * % Passing Reading
    * % Overall Passing (The percentage of students that passed math and reading.)
    * Create a dataframe to hold the above results

In [25]:
# Build School Summary Table
school_summary = pd.DataFrame({'School Name': schools,
                               'School Type': type_of_school,
                               'Total Students': school_total_students,
                               'Total School Budget': budget_by_school,
                               'Per Student Budget': budget_per_student_by_school,
                               'Average Math Score': school_average_math,
                               'Average Reading Score': school_average_reading,
                               '% Passing Math': percent_math_by_school,
                               '% Passing Reading': percent_reading_by_school,
                               '% Overall Passing': percent_both_by_school})

#school_summary

In [26]:
# Use School Name column as index and sort alphabetically
school_summary = school_summary.sort_values('School Name')

school_summary = school_summary.set_index('School Name')

#school_summary

In [27]:
# Create a copy and format to make School Summary Table more readable and accurately present data
school_summary_final = school_summary.copy()

# Add a thousand separator comma to Total Students column
school_summary_final['Total Students'] = school_summary['Total Students'].map("{:,}".format)

# Add a thousand separator comma (if necessary), dollar sign, and decimal values to represents cents to two Budget columns
school_summary_final['Total School Budget'] = school_summary['Total School Budget'].map("${:,.2f}".format)
school_summary_final['Per Student Budget'] = school_summary['Per Student Budget'].map("${:,.2f}".format)

# Only show two decimal places on Average Math Score and Average Reading Score columns
school_summary_final['Average Math Score'] = school_summary['Average Math Score'].map("{:.2f}".format)
school_summary_final['Average Reading Score'] = school_summary['Average Reading Score'].map("{:.2f}".format)

# Only show two decimal places and show percent symbol on Percent Passing columns
school_summary_final['% Passing Math'] = school_summary['% Passing Math'].map("{:.2f}%".format)
school_summary_final['% Passing Reading'] = school_summary['% Passing Reading'].map("{:.2f}%".format)
school_summary_final['% Overall Passing'] = school_summary['% Overall Passing'].map("{:.2f}%".format)

school_summary_final

School Type Total Students Total School Budget  \
School Name                                                            
Bailey High School       District          4,976       $3,124,928.00   
Cabrera High School       Charter          1,858       $1,081,356.00   
Figueroa High School     District          2,949       $1,884,411.00   
Ford High School         District          2,739       $1,763,916.00   
Griffin High School       Charter          1,468         $917,500.00   
Hernandez High School    District          4,635       $3,022,020.00   
Holden High School        Charter            427         $248,087.00   
Huang High School        District          2,917       $1,910,635.00   
Johnson High School      District          4,761       $3,094,650.00   
Pena High School          Charter            962         $585,858.00   
Rodriguez High School    District          3,999       $2,547,363.00   
Shelton High School       Charter          1,761       $1,056,600.00   
Thomas High School        Charter          1,635       $1,043,130.00   
Wilson High School        Charter          2,283       $1,319,574.00   
Wright High School        Charter          1,800       $1,049,400.00   

                      Per Student Budget Average Math Score  \
School Name                                                   
Bailey High School               $628.00              77.05   
Cabrera High School              $582.00              83.06   
Figueroa High School             $639.00              76.71   
Ford High School                 $644.00              77.10   
Griffin High School              $625.00              83.35   
Hernandez High School            $652.00              77.29   
Holden High School               $581.00              83.80   
Huang High School                $655.00              76.63   
Johnson High School              $650.00              77.07   
Pena High School                 $609.00              83.84   
Rodriguez High School            $637.00              76.84   
Shelton High School              $600.00              83.36   
Thomas High School               $638.00              83.42   
Wilson High School               $578.00              83.27   
Wright High School               $583.00              83.68   

                      Average Reading Score % Passing Math % Passing Reading  \
School Name                                                                    
Bailey High School                    81.03         66.68%            81.93%   
Cabrera High School                   83.98         94.13%            97.04%   
Figueroa High School                  81.16         65.99%            80.74%   
Ford High School                      80.75         68.31%            79.30%   
Griffin High School                   83.82         93.39%            97.14%   
Hernandez High School                 80.93         66.75%            80.86%   
Holden High School                    83.81         92.51%            96.25%   
Huang High School                     81.18         65.68%            81.32%   
Johnson High School                   80.97         66.06%            81.22%   
Pena High School                      84.04         94.59%            95.95%   
Rodriguez High School                 80.74         66.37%            80.22%   
Shelton High School                   83.73         93.87%            95.85%   
Thomas High School                    83.85         93.27%            97.31%   
Wilson High School                    83.99         93.87%            96.54%   
Wright High School                    83.95         93.33%            96.61%   

                      % Overall Passing  
School Name                              
Bailey High School               54.64%  
Cabrera High School              91.33%  
Figueroa High School             53.20%  
Ford High School                 54.29%  
Griffin High School              90.60%  
Hernandez High School            53.53%  
Holden High School          

<b>Top Performing Schools (By % Overall Passing)
    
 * Sort and display the top five performing schools by % overall passing.

In [28]:
# Sort data by Overall Passing to find top performing schools
top_performing = school_summary_final.sort_values('% Overall Passing', ascending=False)

# Display top 5 performing schools
top_performing.head()

School Type Total Students Total School Budget  \
School Name                                                          
Cabrera High School     Charter          1,858       $1,081,356.00   
Thomas High School      Charter          1,635       $1,043,130.00   
Griffin High School     Charter          1,468         $917,500.00   
Wilson High School      Charter          2,283       $1,319,574.00   
Pena High School        Charter            962         $585,858.00   

                    Per Student Budget Average Math Score  \
School Name                                                 
Cabrera High School            $582.00              83.06   
Thomas High School             $638.00              83.42   
Griffin High School            $625.00              83.35   
Wilson High School             $578.00              83.27   
Pena High School               $609.00              83.84   

                    Average Reading Score % Passing Math % Passing Reading  \
School Name                                                                  
Cabrera High School                 83.98         94.13%            97.04%   
Thomas High School                  83.85         93.27%            97.31%   
Griffin High School                 83.82         93.39%            97.14%   
Wilson High School                  83.99         93.87%            96.54%   
Pena High School                    84.04         94.59%            95.95%   

                    % Overall Passing  
School Name                            
Cabrera High School            91.33%  
Thomas High School             90.95%  
Griffin High School            90.60%  
Wilson High School             90.58%  
Pena High School               90.54%

<b>Bottom Performing Schools (By % Overall Passing)¶
* Sort and display the five worst-performing schools by % overall passing.

In [29]:
# Sort data by Overall Passing to find bottom performing schools
bottom_performing = school_summary_final.sort_values('% Overall Passing')

# Display bottom 5 performing schools
bottom_performing.head()

School Type Total Students Total School Budget  \
School Name                                                            
Rodriguez High School    District          3,999       $2,547,363.00   
Figueroa High School     District          2,949       $1,884,411.00   
Huang High School        District          2,917       $1,910,635.00   
Hernandez High School    District          4,635       $3,022,020.00   
Johnson High School      District          4,761       $3,094,650.00   

                      Per Student Budget Average Math Score  \
School Name                                                   
Rodriguez High School            $637.00              76.84   
Figueroa High School             $639.00              76.71   
Huang High School                $655.00              76.63   
Hernandez High School            $652.00              77.29   
Johnson High School              $650.00              77.07   

                      Average Reading Score % Passing Math % Passing Reading  \
School Name                                                                    
Rodriguez High School                 80.74         66.37%            80.22%   
Figueroa High School                  81.16         65.99%            80.74%   
Huang High School                     81.18         65.68%            81.32%   
Hernandez High School                 80.93         66.75%            80.86%   
Johnson High School                   80.97         66.06%            81.22%   

                      % Overall Passing  
School Name                              
Rodriguez High School            52.99%  
Figueroa High School             53.20%  
Huang High School                53.51%  
Hernandez High School            53.53%  
Johnson High School              53.54%

In [30]:
# Math Scores By Grade & Reading Scores By Grade
# Create a series for each grade
grade9 = school_data_complete.loc[school_data_complete['grade'] == '9th', :]
grade10 = school_data_complete.loc[school_data_complete['grade'] == '10th', :]
grade11 = school_data_complete.loc[school_data_complete['grade'] == '11th', :]
grade12 = school_data_complete.loc[school_data_complete['grade'] == '12th', :]
#grade9.head()

In [31]:
# GroupBy tables to isolate by school for calculation
grade9_by_school = grade9.groupby(['School ID'])
grade10_by_school = grade10.groupby(['School ID'])
grade11_by_school = grade11.groupby(['School ID'])
grade12_by_school = grade12.groupby(['School ID'])

# Perform calculation to get average math grade by school
math_avg_9th = grade9_by_school['math_score'].mean().round(2)
math_avg_10th = grade10_by_school['math_score'].mean().round(2)
math_avg_11th = grade11_by_school['math_score'].mean().round(2)
math_avg_12th = grade12_by_school['math_score'].mean().round(2)

# Perform calculation to get average reading grade by school
reading_avg_9th = grade9_by_school['reading_score'].mean().round(2)
reading_avg_10th = grade10_by_school['reading_score'].mean().round(2)
reading_avg_11th = grade11_by_school['reading_score'].mean().round(2)
reading_avg_12th = grade12_by_school['reading_score'].mean().round(2)

#math_avg_9th

<b>Math Scores by Grade
* Create a table that lists the average Reading Score for students of each grade level (9th, 10th, 11th, 12th) at each school.
* Create a pandas series for each grade. Hint: use a conditional statement.
* Group each series by school
* Combine the series into a dataframe
* Optional: give the displayed data cleaner formatting

In [32]:
# Build DataFrame for Math Scores By Grade
math_scores_by_grade = pd.DataFrame({'School Name': schools,
                                     '9th': math_avg_9th,
                                     '10th': math_avg_10th,
                                     '11th': math_avg_11th,
                                     '12th': math_avg_12th})

# Use School Name column as index and sort alphabetically
math_scores_by_grade = math_scores_by_grade.sort_values('School Name')
math_scores_by_grade = math_scores_by_grade.set_index('School Name')

math_scores_by_grade

9th   10th   11th   12th
School Name                                      
Bailey High School     77.08  77.00  77.52  76.49
Cabrera High School    83.09  83.15  82.77  83.28
Figueroa High School   76.40  76.54  76.88  77.15
Ford High School       77.36  77.67  76.92  76.18
Griffin High School    82.04  84.23  83.84  83.36
Hernandez High School  77.44  77.34  77.14  77.19
Holden High School     83.79  83.43  85.00  82.86
Huang High School      77.03  75.91  76.45  77.23
Johnson High School    77.19  76.69  77.49  76.86
Pena High School       83.63  83.37  84.33  84.12
Rodriguez High School  76.86  76.61  76.40  77.69
Shelton High School    83.42  82.92  83.38  83.78
Thomas High School     83.59  83.09  83.50  83.50
Wilson High School     83.09  83.72  83.20  83.04
Wright High School     83.26  84.01  83.84  83.64

<b>Reading Score by Grade
* Perform the same operations as above for reading scores

In [33]:
# Build DataFrame for Reading Scores By Grade
reading_scores_by_grade = pd.DataFrame({'School Name': schools,
                                     '9th': reading_avg_9th,
                                     '10th': reading_avg_10th,
                                     '11th': reading_avg_11th,
                                     '12th': reading_avg_12th})

# Use School Name column as index and sort alphabetically
reading_scores_by_grade = reading_scores_by_grade.sort_values('School Name')
reading_scores_by_grade = reading_scores_by_grade.set_index('School Name')

reading_scores_by_grade

9th   10th   11th   12th
School Name                                      
Bailey High School     81.30  80.91  80.95  80.91
Cabrera High School    83.68  84.25  83.79  84.29
Figueroa High School   81.20  81.41  80.64  81.38
Ford High School       80.63  81.26  80.40  80.66
Griffin High School    83.37  83.71  84.29  84.01
Hernandez High School  80.87  80.66  81.40  80.86
Holden High School     83.68  83.32  83.82  84.70
Huang High School      81.29  81.51  81.42  80.31
Johnson High School    81.26  80.77  80.62  81.23
Pena High School       83.81  83.61  84.34  84.59
Rodriguez High School  80.99  80.63  80.86  80.38
Shelton High School    84.12  83.44  84.37  82.78
Thomas High School     83.73  84.25  83.59  83.83
Wilson High School     83.94  84.02  83.76  84.32
Wright High School     83.83  83.81  84.16  84.07

In [34]:
# Scores by School Spending
# Create spending bins
spending_bins = [0, 585, 630, 645, 680]

# Create names for the spendings bins
spending_ranges = ['<$585', '$585-630', '$630-645', '$645-680']

# Add column and assign spending ranges to school_summary table
# Must use unformatted version of table so the budget column we are cutting is still a float datatype
school_summary['Spending Ranges (Per Student)'] = pd.cut(school_summary["Per Student Budget"], spending_bins, labels=spending_ranges, include_lowest=True)

#school_summary.head()

<b> Scores by School Spending
* Create a table that breaks down school performances based on average Spending Ranges (Per Student). Use 4 reasonable bins to group school spending. Include in the table each of the following:
* Average Math Score
* Average Reading Score
* % Passing Math
* % Passing Reading
* Overall Passing Rate (Average of the above two)

In [35]:
# Use GroupBy to filter by bins
school_spending_table = school_summary.groupby('Spending Ranges (Per Student)')

# List of columns to evaluate and include in table
by_spending_columns = ['Average Math Score', 'Average Reading Score', '% Passing Math', '% Passing Reading', '% Overall Passing']

# Create table with average values per spending range
scores_by_spending = school_spending_table[by_spending_columns].mean().round(2)

# Show percent symbol on Percent Passing columns
scores_by_spending['% Passing Math'] = scores_by_spending['% Passing Math'].map("{:}%".format)
scores_by_spending['% Passing Reading'] = scores_by_spending['% Passing Reading'].map("{:}%".format)
scores_by_spending['% Overall Passing'] = scores_by_spending['% Overall Passing'].map("{:}%".format)

scores_by_spending

Average Math Score  Average Reading Score  \
Spending Ranges (Per Student)                                              
<$585                                       83.46                  83.93   
$585-630                                    81.90                  83.16   
$630-645                                    78.52                  81.62   
$645-680                                    77.00                  81.03   

                              % Passing Math % Passing Reading  \
Spending Ranges (Per Student)                                    
<$585                                 93.46%            96.61%   
$585-630                              87.13%            92.72%   
$630-645                              73.48%            84.39%   
$645-680                              66.16%            81.13%   

                              % Overall Passing  
Spending Ranges (Per Student)                    
<$585                                    90.37%  
$585-630                                 81.42%  
$630-645                                 62.86%  
$645-680                                 53.53%

In [36]:
# Scores by School Size
# Create size bins
size_bins = [0, 1000, 2000, 5000]

# Create names for the spendings bins
size_ranges = ['Small (<1000)', 'Medium (1000-2000)', 'Large (2000-5000)']

# Add column and assign spending ranges to school_size table
# Must use unformatted version of table so the total students column we are cutting is still an int datatype
school_summary['School Size'] = pd.cut(school_summary["Total Students"], size_bins, labels=size_ranges, include_lowest=True)

#school_summary.head()

<b>Scores by School Size
* Perform the same operations as above, based on school size.

In [37]:
# Use GroupBy to filter by school size bins
school_size_table = school_summary.groupby('School Size')

# List of columns to evaluate and include in table
by_size_columns = ['Average Math Score', 'Average Reading Score', '% Passing Math', '% Passing Reading', '% Overall Passing']

# Create table with average values per size range
scores_by_size = school_size_table[by_size_columns].mean().round(2)

# Show percent symbol on Percent Passing columns
scores_by_size['% Passing Math'] = scores_by_size['% Passing Math'].map("{:}%".format)
scores_by_size['% Passing Reading'] = scores_by_size['% Passing Reading'].map("{:}%".format)
scores_by_size['% Overall Passing'] = scores_by_size['% Overall Passing'].map("{:}%".format)

scores_by_size

Average Math Score  Average Reading Score % Passing Math  \
School Size                                                                    
Small (<1000)                    83.82                  83.93         93.55%   
Medium (1000-2000)               83.37                  83.86          93.6%   
Large (2000-5000)                77.75                  81.34         69.96%   

                   % Passing Reading % Overall Passing  
School Size                                             
Small (<1000)                  96.1%            89.88%  
Medium (1000-2000)            96.79%            90.62%  
Large (2000-5000)             82.77%            58.29%

<b>Scores by Scool Type
* Perform the same operations as above, based on school size.

In [38]:
# Use GroupBy to filter by school type
school_type_table = school_summary.groupby('School Type')

# List of columns to evaluate and include in table
by_type_columns = ['Average Math Score', 'Average Reading Score', '% Passing Math', '% Passing Reading', '% Overall Passing']

# Create table with average values per size range
scores_by_type = school_type_table[by_type_columns].mean().round(2)

# Show percent symbol on Percent Passing columns
scores_by_type['% Passing Math'] = scores_by_type['% Passing Math'].map("{:}%".format)
scores_by_type['% Passing Reading'] = scores_by_type['% Passing Reading'].map("{:}%".format)
scores_by_type['% Overall Passing'] = scores_by_type['% Overall Passing'].map("{:}%".format)


scores_by_type

Average Math Score  Average Reading Score % Passing Math  \
School Type                                                             
Charter                   83.47                  83.90         93.62%   
District                  76.96                  80.97         66.55%   

            % Passing Reading % Overall Passing  
School Type                                      
Charter                96.59%            90.43%  
District                80.8%            53.67%

## PyCity Schools Report

The Disrict Summary Table showed that as a whole, students in the district had a higher pass rate in reading than math by over 10 percentage points (85.81% vs 74.98%).

The School Summary Table and the filtered versions showing top and bottom performing showed even more insight on how schools performed compared to one another.
* The Top 5 performing schools are all Charter schools
* The Bottom 5 performing schools are all District schools
* Aside from the 5 Charter schools performing the best, they also are doing it with lower per student budgets (on average) than the District schools in the bottom 5.

The tables for Reading and Math Scores by Grades showed that scores stayed relatively consistent at each school from 9th to 12th grade.

The Scores by School Spending confirmed an earlier observation that the schools spending less per student are actually outperforming the schools spending more per student. In fact, the average Overall Pass % steadily decreased from over 90% for the lowest range (<$585 per student) to less than 54% for the highest range ($645 to $680 per student).

The Scores by School Size table shows that schools designated to be Small (<1000 students) or Medium (1000-2000 students) both have similar average pass rates around 90% while large schools (2000-5000 students) dropped steeply to an average pass rate around 58%.

The Scores by School Type table reinforced a previous observation that Charter schools outperform District schools with average pass rates of approximately 90% and 54% respectively.